py/dml/dmlc.py

```diff
@@ -153,6 +154,60 @@ def print_cdep(outputbase, headers, footers):
     f.close()
     f.commit()
 
+def dotdot_depth(path):
+    '''return how many levels of parent directories are required to resolve a
+    relative path'''
+    return os.path.normpath(path).count('..' + os.path.sep)
+
+assert dotdot_depth('./../../foo/../../../bar.dml') == 4
+
+def dump_input_files(outputbase, imported):
+    # DML is often invoked within a complex build system, with a large
+    # number of -I flags and many auto-generated files. It is often
+    # hard to reproduce a DML issue outside of this build system. This
+    # function tries to save all relevant DML files in a tree that is
+    # buildable or near-buildable, by placing all DML files in the
+    # same directory. For each files that is imported with a relative
+    # path, a back-pointing symlink is created on that relative path.
+    max_dotdot = max(dotdot_depth(p)
+                     for (_, paths) in imported.items()
+                     for p in paths)
+    prefix = '/'.join(['_'] * max_dotdot)
+    basenames = set()
+    with tarfile.open(outputbase + ".tar.xz", 'w:xz') as tf:
+        # if two different files foo.dml are imported, where one is
+        # imported as "foo.dml" or "./foo.dml" and the other is only
+        # imported using a qualified path such as "bar/foo.dml", then
+        # the rename hack below has a better chance to work if we process
+        # the unqualified paths first.
+        for f in sorted(imported, key=lambda f: min(map(len, imported[f]))):
```

Comment:
This is a very hacky hack. I'd like to see a better approach. If you can't think of one, add a `HACK` annotation.

Reply:
My approach is inherently unsafe, so there is no natural clean solution, just a risk-minimizing heuristics. So I'll go for `# HACK` (the code probably deserves more than one such annotation).